### ISSUE TITLE
IDA backend: include api features of renamed calls

### ISSUE THREAD

**Issue by mr-tz** [mandiant/capa#252]
### Summary

After renaming dynamic calls in IDA, the backend should emit the respective API calls.

Most often the dynamic calls will be to global addresses, but calls via registers or even local variables are also possible.

### Motivation

Malware often contains obfuscated API calls as shown below.

Before
![2020-08-27_10-06-25](https://user-images.githubusercontent.com/17606537/91414641-22ab5100-e84d-11ea-897a-b67096fdd3d4.png)

After
![2020-08-27_10-06-32](https://user-images.githubusercontent.com/17606537/91414646-24751480-e84d-11ea-9226-f325d2e023cd.png)

### Describe alternatives you've considered

Alternatively, an additional analysis engine could try to automatically recover/identify API calls, e.g. using emulation. This could then work in the standalone version as well.

**Comment by mike-hunhoff** [comment]
+1 for this feature. I'm analyzing some complicated shellcode that dynamically resolves all API calls and stores the addresses in a large structure. We should investigate solutions to include these sorts of API calls as well.

**Comment by mike-hunhoff** [comment]
maybe can adapt code here -> https://github.com/arizvisa/ida-minsc/blob/master/base/instruction.py#L819 to extract API calls from user-defined structures.

**Comment by mike-hunhoff** [comment]
After doing additional research it appears that attempting to pull API names from a user-defined structure using IDAPython can get messy real quick - especially when working with large, nested structures.

A simpler solution may be to check if a `call`/`jmp` references a structure offset using something like `idaapi.is_stroff` and then parse the structure member using a regular expression, something like `(call|jmp)\s+\[.+\s*\+\s*(.+)\]`.

I don't like the idea of parsing the disassembly like this but it's the simplest solution and may require less overhead then attempting to go through IDAPython.

**Comment by mike-hunhoff** [comment]
Another problem to tackle - how do we map something like `GetMessageW` to `user32.GetMessageW`? Most capa rules specify the DLL name as part of the rule which means we can't match with just `GetMessageW`.

One option is to only support specific annotation formats e.g. `WIN.api.user32_GetMessageW`

**Comment by Ana06** [comment]
> Another problem to tackle - how do we map something like GetMessageW to user32.GetMessageW? Most capa rules specify the DLL name as part of the rule which means we can't match with just GetMessageW.

I think in the extractor we only use the API name: https://github.com/mandiant/capa/blob/bb86d1485c0d9633ccef5431cd95fca5e709b20c/capa/features/extractors/ida/insn.py#L83 So I assume the DLL name in the rules is documentation and not really used.

**Comment by Ana06** [comment]
Adding support for calls to global variables can be implemented by getting the name of the `call`/`jump` operand. I'll send a PR to support it. :wink:

**Comment by Ana06** [comment]
@mike-hunhoff 
> A simpler solution may be to check if a call/jmp references a structure offset using something like idaapi.is_stroff and then parse the structure member using a regular expression, something like (call|jmp)\s+\[.+\s*\+\s*(.+)\].

I think this could work. But I understand this would need that the structure has been applied in the disassembly. If the pseudocode view work well, I often apply the structures only in the pseudocode view and it wouldn't work in that case. So not sure if the effort of implementing it is worth it.

Can you share the hash of a sample that uses this?

**Comment by Ana06** [comment]
@mike-hunhoff 
> I'm analyzing some shellcode that dynamically resolves all API calls and stores the addresses in a large structure. 

Can you share the hash of the sample?

> A simpler solution may be to check if a call/jmp references a structure offset using something like idaapi.is_stroff and then parse the structure member using a regular expression, something like (call|jmp)\s+\[.+\s*\+\s*(.+)\].

I think this could work. But I understand this would need that the structure has been applied in the disassembly. If the pseudocode view work well, I often apply the structures only in the pseudocode view and it wouldn't work in that case. We could also parse the pseudocode view with code like the following which returns the pseudocode line corresponding to the `call` instruction:
```python
cfunc = ida_hexrays.decompile(ea)
item = cfunc.body.find_closest_addr(ea)
coord = cfunc.find_item_coords(item)
cfunc.get_pseudocode()[coord[1]].line 
```

But I am not sure how efficient parsing the pseudocode is. We likely would want to ensure we only decompile every function once.

**Comment by Ana06** [comment]
Registers are also tricky, as they are often the value returned by a function that resolves the APIs (and this function could be very different). An example of a sample I analysed yesterday:
![image](https://github.com/mandiant/capa/assets/16052290/2dcbcf66-a39c-4ac7-947d-10abb8bdae6b)

I often add a comment in the call instruction with the API. Not sure if this is a standard practice, but if it is, getting the API from the comment should be easy. Parsing the pseudocode could also be used for this case.

**Comment by Ana06** [comment]
Does someone has a sample where the resolved APIs are stored in local variables? I think this case could also be tricky if the variable is reused.

**Comment by mr-tz** [comment]
closed via #2201
